File: packages/native/tests/multi_paint_test.cc
// Licensed under the Apache License Version 2.0 that may not be used except in
// compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.
//
// Host-side unit tests for the multi-<Paint> multi-pass channel (FEATURE_PARITY
// F.1.3 / roadmap #13): SetMultiPaint installs a full pass list on the retained
// node (nested MultiPaintList bytes, JS-built on the platform side — rebuilt
// here with the same FlatBuffer builder), each pass carrying fully independent
// paint state; an empty payload clears the list.

#include "../shared/skity/retained_render_tree.h"

#include <vector>

#include <flatbuffers/flatbuffers.h>
#include <gtest/gtest.h>

#include "command_batch_generated.h"

namespace {

using namespace skityrt;

// Batch builder helpers (same pattern as retained_tree_version_test.cc —
// offsets are builder-local, so each command must be created on the SAME
// builder as the batch that references it).
class BatchBuilder {
public:
  flatbuffers::FlatBufferBuilder fbb;
  std::vector<flatbuffers::Offset<void>> cmds;
  std::vector<uint8_t> types;

  template <typename Fn> void Add(Command type, Fn make) {
    cmds.push_back(make(fbb).Union());
    types.push_back(static_cast<uint8_t>(type));
  }

  std::vector<uint8_t> Finish() {
    auto batch =
        skityrt::CreateCommandBatch(fbb, 0, fbb.CreateVector(types), fbb.CreateVector(cmds));
    skityrt::FinishCommandBatchBuffer(fbb, batch);
    return std::vector<uint8_t>(fbb.GetBufferPointer(), fbb.GetBufferPointer() + fbb.GetSize());
  }
};

std::vector<uint8_t> MakeTree() {
  BatchBuilder b;
  b.Add(Command_InsertNode, [](flatbuffers::FlatBufferBuilder &f) {
    return skityrt::CreateInsertNode(f, 1, -1, 0, f.CreateString("canvas"));
  });
  b.Add(Command_InsertNode, [](flatbuffers::FlatBufferBuilder &f) {
    return skityrt::CreateInsertNode(f, 2, 1, 0, f.CreateString("rect"));
  });
  return b.Finish();
}

// A nested MultiPaintList blob: a red fill pass (half opacity, plus blend
// mode) followed by a wide white stroke pass with dashes.
std::vector<uint8_t> MakeMultiPaintListBlob() {
  flatbuffers::FlatBufferBuilder f;
  auto dash = f.CreateVector(std::vector<float>{6.f, 4.f});
  PaintPassBuilder p0(f);
  p0.add_style(PaintSlot_FILL);
  p0.add_type(1); // COLOR
  p0.add_color(0xffff0000);
  p0.add_opacity(0.5f);
  p0.add_blend_mode(BlendMode_MULTIPLY);
  auto off0 = p0.Finish();
  PaintPassBuilder p1(f);
  p1.add_style(PaintSlot_STROKE);
  p1.add_type(1);
  p1.add_color(0xffffffff);
  p1.add_stroke_width(3.f);
  p1.add_stroke_cap(LineCap_ROUND);
  p1.add_opacity(1.f);
  p1.add_stroke_dash(dash);
  auto off1 = p1.Finish();
  auto passes = f.CreateVector(std::vector<flatbuffers::Offset<PaintPass>>{off0, off1});
  MultiPaintListBuilder list(f);
  list.add_passes(passes);
  f.Finish(list.Finish());
  return std::vector<uint8_t>(f.GetBufferPointer(), f.GetBufferPointer() + f.GetSize());
}

std::vector<uint8_t> MakeSetMultiPaint(int32_t node, const std::vector<uint8_t> &blob) {
  BatchBuilder b;
  b.Add(Command_SetMultiPaint, [&](flatbuffers::FlatBufferBuilder &f) {
    auto data = f.CreateVector(blob);
    return skityrt::CreateSetMultiPaint(f, node, data);
  });
  return b.Finish();
}

} // namespace

// @lat: [[tests#Native C++ core#Multi-pass paint]]
TEST(MultiPaintTest, InstallsPassesWithIndependentState) {
  RetainedRenderTree tree;
  auto insert = MakeTree();
  tree.ApplyCommandBatch(insert.data(), insert.size());
  const RetainedNode *node = tree.Find(2);
  ASSERT_NE(node, nullptr);
  EXPECT_TRUE(node->multi_passes.empty());
  auto version = node->paint_version;

  auto blob = MakeMultiPaintListBlob();
  auto batch = MakeSetMultiPaint(2, blob);
  tree.ApplyCommandBatch(batch.data(), batch.size());

  ASSERT_EQ(node->multi_passes.size(), 2u);
  // Pass 0: red fill, half opacity, multiply blend — the fields the
  // single-slot channel shares, made per-pass.
  const RetainedPaintPass &fill = node->multi_passes[0];
  EXPECT_FALSE(fill.stroke);
  EXPECT_EQ(fill.paint.type, 1u);
  EXPECT_EQ(fill.paint.color, 0xffff0000u);
  EXPECT_FLOAT_EQ(fill.opacity, 0.5f);
  EXPECT_EQ(fill.blend_mode, BlendMode_MULTIPLY);
  // Pass 1: white stroke, own width/cap/dash — independent stroke attrs.
  const RetainedPaintPass &stroke = node->multi_passes[1];
  EXPECT_TRUE(stroke.stroke);
  EXPECT_EQ(stroke.paint.type, 1u);
  EXPECT_EQ(stroke.paint.color, 0xffffffffu);
  EXPECT_FLOAT_EQ(stroke.stroke_width, 3.f);
  EXPECT_EQ(stroke.stroke_cap, LineCap_ROUND);
  ASSERT_NE(stroke.stroke_dash, nullptr);
  ASSERT_EQ(stroke.stroke_dash->size(), 2u);
  EXPECT_FLOAT_EQ((*stroke.stroke_dash)[0], 6.f);
  EXPECT_FLOAT_EQ((*stroke.stroke_dash)[1], 4.f);
  // Full-state write bumps the paint build-cache version.
  EXPECT_GT(node->paint_version, version);
}

// @lat: [[tests#Native C++ core#Multi-pass paint]]
TEST(MultiPaintTest, FullStateReplaceAndClear) {
  RetainedRenderTree tree;
  auto insert = MakeTree();
  tree.ApplyCommandBatch(insert.data(), insert.size());
  const RetainedNode *node = tree.Find(2);
  ASSERT_NE(node, nullptr);

  auto blob = MakeMultiPaintListBlob();
  auto batch = MakeSetMultiPaint(2, blob);
  tree.ApplyCommandBatch(batch.data(), batch.size());
  ASSERT_EQ(node->multi_passes.size(), 2u);

  // A second command REPLACES the whole list (full state, not append).
  std::vector<uint8_t> one;
  {
    flatbuffers::FlatBufferBuilder f;
    PaintPassBuilder p(f);
    p.add_style(PaintSlot_FILL);
    p.add_type(1);
    p.add_color(0xff00ff00);
    auto off = p.Finish();
    auto passes = f.CreateVector(std::vector<flatbuffers::Offset<PaintPass>>{off});
    MultiPaintListBuilder list(f);
    list.add_passes(passes);
    f.Finish(list.Finish());
    one.assign(f.GetBufferPointer(), f.GetBufferPointer() + f.GetSize());
  }
  auto replace = MakeSetMultiPaint(2, one);
  tree.ApplyCommandBatch(replace.data(), replace.size());
  ASSERT_EQ(node->multi_passes.size(), 1u);
  EXPECT_EQ(node->multi_passes[0].paint.color, 0xff00ff00u);

  // An empty payload clears — the node falls back to its single-slot paints.
  auto clear = MakeSetMultiPaint(2, {});
  tree.ApplyCommandBatch(clear.data(), clear.size());
  EXPECT_TRUE(node->multi_passes.empty());
}

// @lat: [[tests#Native C++ core#Multi-pass paint]]
TEST(MultiPaintTest, UnknownNodeIsIgnored) {
  RetainedRenderTree tree;
  auto insert = MakeTree();
  tree.ApplyCommandBatch(insert.data(), insert.size());
  auto blob = MakeMultiPaintListBlob();
  auto batch = MakeSetMultiPaint(99, blob);
  tree.ApplyCommandBatch(batch.data(), batch.size()); // no crash, no-op
  EXPECT_EQ(tree.Find(99), nullptr);
}

File: packages/graphics/src/__tests__/multi-paint.test.ts
import { describe, expect, it } from "vitest";

import { buildMultiPaint } from "../multi-paint.js";
import { MultiPaintList } from "../generated/skityrt/multi-paint-list.js";
import { BlendMode as BM } from "../generated/skityrt/blend-mode.js";
import { LineCap } from "../generated/skityrt/line-cap.js";
import { PaintSlot } from "../generated/skityrt/paint-slot.js";
import * as flatbuffers from "../generated/flatbuffers/flatbuffers.js";

function readBack(bytes: ArrayBuffer): MultiPaintList {
  return MultiPaintList.getRootAsMultiPaintList(new flatbuffers.ByteBuffer(new Uint8Array(bytes)));
}

// @lat: [[tests#Graphics parsing layer#Multi-paint blob]]
describe("buildMultiPaint", () => {
  it("returns null for an empty pass list", () => {
    expect(buildMultiPaint([])).toBeNull();
  });

  it("serializes each pass with independent state, in order", () => {
    const bytes = buildMultiPaint([
      { style: "fill", color: 0xffff0000, opacity: 0.5, blendMode: BM.MULTIPLY },
      {
        style: "stroke",
        color: 0xffffffff,
        strokeWidth: 3,
        strokeCap: LineCap.ROUND,
        dash: [6, 4],
      },
    ]);
    expect(bytes).not.toBeNull();
    const list = readBack(bytes!);
    expect(list.passesLength()).toBe(2);

    const p0 = list.passes(0)!;
    expect(p0.style()).toBe(PaintSlot.FILL);
    expect(p0.type()).toBe(1); // COLOR
    expect(p0.color()).toBe(0xffff0000);
    expect(p0.opacity()).toBe(0.5);
    expect(p0.blendMode()).toBe(BM.MULTIPLY);

    const p1 = list.passes(1)!;
    expect(p1.style()).toBe(PaintSlot.STROKE);
    expect(p1.color()).toBe(0xffffffff);
    expect(p1.strokeWidth()).toBe(3);
    expect(p1.strokeCap()).toBe(LineCap.ROUND);
    expect(p1.strokeDashLength()).toBe(2);
    expect(p1.strokeDash(0)).toBe(6);
    expect(p1.strokeDash(1)).toBe(4);
  });

  it("round-trips gradient and filter bytes verbatim", () => {
    const gradient = new Uint8Array([1, 2, 3, 4]);
    const colorFilter = new Uint8Array([9, 8, 7]);
    const bytes = buildMultiPaint([
      {
        style: "fill",
        gradient: gradient.slice().buffer,
        colorFilter: colorFilter.slice().buffer,
      },
    ]);
    const list = readBack(bytes!);
    const p = list.passes(0)!;
    expect(p.type()).toBe(2); // GRADIENT
    expect(Array.from(p.gradientArray() ?? [])).toEqual([1, 2, 3, 4]);
    expect(Array.from(p.colorFilterArray() ?? [])).toEqual([9, 8, 7]);
  });

  it("serializes the image shader fields", () => {
    const bytes = buildMultiPaint([
      {
        style: "fill",
        image: { uri: "https://x/y.png", fit: 2, tx: 1, ty: 2, rect: [1, 2, 3, 4] },
      },
    ]);
    const list = readBack(bytes!);
    const p = list.passes(0)!;
    expect(p.type()).toBe(3); // IMAGE_SHADER
    expect(p.imageUri()).toBe("https://x/y.png");
    expect(p.imageFit()).toBe(2);
    expect(p.imageTx()).toBe(1);
    expect(p.imageTy()).toBe(2);
    expect(p.imageRectLength()).toBe(4);
  });

  it("an inactive pass (no paint source) serializes with type NONE", () => {
    const bytes = buildMultiPaint([{ style: "fill" }]);
    const list = readBack(bytes!);
    expect(list.passes(0)!.type()).toBe(0);
  });
});
